### PR TITLE
fix issue 115 allow setting ydl_options

### PR DESCRIPTION
**Pull request by kkrolczyk**
Pafy had been changed minimally, to make sure it's impossible to introduce any errors. It was possible to add global function similar to "set_api_key" which would modify g.ydl_opts, however, this would require following of and modifying object (pafy) creation code in mps-yt. Different patchset for mps uses new ydl_opts.

### CODE REVIEW

**Comment by ids1024** [comment]
This looks good overall, and thanks for contributing this and the change for mpsyt.

One thing. I think it would be better if instead of completely overriding the defaults set in pafy, it could override only the keys specified. For example, `quiet` will be set to `True`, unless the caller specifically specifies it as `False`. This could be done by making a copy of the default `ytdl_options` dictionary and using `dict.update()` with the dictionary specified as an argument to `new()`.

What do you think of this?

**Comment by kkrolczyk** [comment]
@ids1024 thanks for feedback - sure, I think it's a good idea, yet raises tiny question about architecture. Currently it is mpsyt who requests for pafy instance which in turn (please correct me if i'm wrong) is somewhat wrapper for youtube-dl. And also - separate configurations are kept in pafy as well as in mpsyt. I wonder if multitude of different configurations could be somehow reduced/unified.
It's possible to copy pafy's default ydl_opts, update it just to send it back to initialize Pafy - however, I suppose this doesn't sound too good. How about having default ydl_opts settings duplicated in mpsyt 'g' module. I'm not quite fond of code duplication, but I assume You want to maintain backwards compatibility for pafy, so ydl_options couldn't be removed from there. 

tl;dr: how about having mpsyt g.ydl_opts dict initialized to values from pafy default ydl_opts, instead of copying back and forth?

**Comment by ids1024** [comment]
Hm, I'm not sure about the best option.
- What I was saying is that, if a blank options dictionary is passed to `pafy.new()`, it should use it's defaults. It will only, say, change quiet to `False` if the options dictionary passed includes a `'quiet'` key. This does not require mpsyt, or another client, to know what the defaults are. The default `yt_dl_cmdline` option in mpsyt would be blank.
- Another option would be for pafy to just not have defaults, and let mpsyt (or another client handle it). Thus, the youtube-dl options setting in mpsyt would have a value of `--quiet --no-warnings --prefer-insecure`.
- Another option would be like the above, but pafy would have a default that is used if no options dictionary is provided. The default options would be duplicated in both pafy and mpsyt.
- Perhaps something could be done where mpsyt reads the defaults from pafy. This would require changes to pafy, as `__init__.py` imports only a few functions from `pafy.py` and then deletes the reference to it with `del`, so the globals are not accessible. Which makes sense, as it is an implementation detail.

**Comment by kkrolczyk** [comment]
Current pafy defaults on develop branch are:
```python
g.ydl_opts = {"prefer_insecure": True, "no_warnings": True, "quiet": True}
```

#### The question A is, what should pafy do with ydl_opts received from user:

1. completely replace own by those user provided (giving user maximum control over configuration) - or use own defaults if none are given (**previous version on fix branch**), 
  ```python 
  class Pafy(...):
      self._ydl_opts = ydl_opts or g.ydl_opts
  ```

2. merge with own defaults optionally replacing them (in case user provided some options like `{'videopassword':'foo', 'prefer_insecure':False}`, but forgot, or ignored `quiet`),
 final ydl_opts passed would look like
`{'videopassword':'foo', 'prefer_insecure':False, "quiet":True, "no_warnings:"True"}`
(**current version on fix branch**)

3. merge but prevent overriding defaults 
`{'videopassword':'foo', 'prefer_insecure':True, "quiet":True, "no_warnings:"True"}`

+ imho (1) or (2) is best choice. In Your comment You seemed to like 2 so i rewrote it to be such.

2-gives maximum backwards compatibility,
1-gives maximum configurability,
3-is kind of user friendly, but limits configurability. 

2-slight minus is, that it does set things behind user's back, and it would enforce some hacky way for him to unset (not change True=>False, just not specify at all) those defaults, like requiring to pass {"quiet": None} or something like this; however i'm not sure that anyone would want to do that, so it might not be an issue.

#### Question B is - where should default ydl_opts reside:

(and how this affects keeping/setting/displaying those options in mps-yt)

a) in mpsyt only			(Your comment's second bullet),
b) in pafy only				(Your comment's fourth bullet),
c) in both mpsyt and pafy	(Your comment's third bullet)

a - it would break other pafy clients that didn't update,
b - it is possible, given pafy changes to update own defaults with user provided changes,
c - it is redundant, given above change in pafy, yet wouldn't do any harm.

a - default ydl_opts would need to be shown(set) in ydl_opts_cmdline, or hidden as defaults to be updated whenever user sets ydl_opts_cmdline
b - default ydl_opts might be shown(set) in ydl_opts_cmdline (explicit) or not (implicit), to display them, pafy should also have getter function
c - this could bring more confusion to anyone browsing code, why are opts duplicated.

+ I think b) is best solution. Mpsyt user need not to be informed that quiet is set behind it's back yet has possibility to set his own ydl arguments as well as override default ones (even if they are not shown explicitly as being set on ydl_opts_cmdline). Currently ydl_opts_cmdline is empty, I think it does not add to confusion.

### TL;DR: 
  this patch now simply updates g.ydl_opts with user provided to new() function, if there are any provided, if not - uses old (default) ones.

**Comment by ids1024** [comment]
It doesn't seem very clean to me for an argument to `pafy.new()` to not only effect that pafy instance, but all instances created in the future.

**Comment by ids1024** [comment]
Merged, but as stated above I made it specific to the individual pafy instance (b21c37089ef35cb435ca5287453e6e7ff569e593)/
